### PR TITLE
Sampler definition

### PR DESCRIPTION
**Pull request by litherum**
Flesh out the definition of WebGPUSamplerDescriptor

### CODE REVIEW

File: design/sketch.webidl [hunk 1]
```diff
@@ -106,15 +106,54 @@ interface WebGPUTexture {
 };
 
 // Sampler
-enum WebGPUFilterMode {
```
Comment (kainino0x) [inline design/sketch.webidl]: Unintentional removal?

Comment (litherum) [reply design/sketch.webidl]: It's on line 121 below.

**Comment by litherum** [comment]
Yeah, I did the investigation. The only thing that was non-obvious was the bias parameter - all the shading languages allow specifying the bias in the sample() function, but only HLSL and Vulkan allow setting the bias in the sampler. Emulating it using a constant buffer would be pretty easy in Metal, and we're likely going to have to create these invisible constant buffers anyway, so I figure we can just do that.

File: design/sketch.webidl (hunk 1 above)
Comment (kainino0x) [reply design/sketch.webidl]: Oh, in that case, I assume you meant to remove the 3 lines below this?

Comment (litherum) [reply design/sketch.webidl]: ....yep.

File: design/sketch.webidl [hunk 2]
```diff
@@ -105,16 +105,51 @@ interface WebGPUTexture {
...
+    WebGPUddressMode rAddressMode = "clampToEdge",
+    WebGPUddressMode sAddressMode = "clampToEdge",
+    WebGPUddressMode tAddressMode = "clampToEdge",
+    WebGPUFilterModeEnum magFilter = "nearest";
+    WebGPUFilterModeEnum minFilter = "nearest";
+    WebGPUFilterModeEnum mipmapFilter = "nearest";
+    float lodMinClamp = 0,
+    float lodMaxClamp = FLT_MAX,
+    unsigned long maxAnisotropy = 1,
+    boolean normalizedCoordinates = true,
```
Comment (Kangz) [inline design/sketch.webidl]: I didn't see something like this in D3D12_SAMPLER_DESC.

Comment (litherum) [reply design/sketch.webidl]: Yep, you’re right, I missed that. It would have to be emulated, too.

**Comment by litherum** [comment]
We should discuss emulation in the group.

**Comment by kvark** [comment]
> The only thing that was non-obvious was the bias parameter - all the shading languages allow specifying the bias in the sample() function, but only HLSL and Vulkan allow setting the bias in the sampler.

Interestingly, we've been calling `setLodBias` on `MTLSamplerDescriptor` unconditionally and it didn't trigger an "unrecognized selector sent to an instance..." failure.

File: design/sketch.webidl [hunk 3]
```diff
@@ -105,16 +105,51 @@ interface WebGPUTexture {
     WebGPUTextureView createTextureView(WebGPUTextureViewDescriptor desc);
 };
 
-// Sampler
+// Samplers
+enum WebGPUAddressMode {
+    "clampToEdge",
+    "mirrorClampToEdge",
```
Comment (kvark) [inline design/sketch.webidl]: note: this requires "VK_KHR_sampler_mirror_clamp_to_edge". It's not supported on Mali/Adreno Android GPUs, and a few desktop ones (e.g. Kabylake GT2 on Linux). Perhaps we can get started without it?

File: design/sketch.webidl [hunk 4]
```diff
@@ -105,16 +105,51 @@ interface WebGPUTexture {
...
+    WebGPUFilterModeEnum magFilter = "nearest";
+    WebGPUFilterModeEnum minFilter = "nearest";
+    WebGPUFilterModeEnum mipmapFilter = "nearest";
+    float lodMinClamp = 0,
+    float lodMaxClamp = FLT_MAX,
+    unsigned long maxAnisotropy = 1,
+    boolean normalizedCoordinates = true,
+    WebGPUCompareFunction compareFunction = "never",
+    WebGPUBorderColor borderColor = "transparentBlack"
+    float bias = 0;
```
Comment (kvark) [inline design/sketch.webidl]: let's call it `lodBias` for better clarity

File: design/sketch.webidl (hunk 2 above)
Comment (kvark) [reply design/sketch.webidl]: Perhaps, we should instead remove this for now?

**Comment by kainino0x** [comment]
Looks like the WebGPUCompareFunction was actually a duplication, I'm going to remove it in a cleanup commit to make this pass IDL validation again.

**Pull request merged**